mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -36,31 +37,131 @@ namespace mitigation {
+                rewriter.setInsertionPointToStart(currentFnFoldedOpBlock);
+                currentFnFoldedOpBlock->addArgument(currentFnFoldedOp.getArgumentTypes().back(),
+                                                    loc);
+
+                bool containQnodes = false;
+                funcOp.walk([&](func::CallOp op) {
+                    auto insideFuncOp =
+                        SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
+                    if (insideFuncOp->hasAttr("qnode")) {
+                        containQnodes = true;
+                        return WalkResult::interrupt();
+                    }
+                    return WalkResult::advance();
+                });
+
+                // Replace calls with calls.zne
+                if (containQnodes) {
+
+                    currentFnFoldedOp.walk([&](func::CallOp callOp) {
+                        PatternRewriter::InsertionGuard insertGuard(rewriter);
+                        func::FuncOp funcOp = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(
+                            op, callOp.getCalleeAttr());
+                        std::string foldedName = callOp.getCalleeAttrName().str() + ".folded";
+                        func::FuncOp foldedOp = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(
+                            moduleOp, rewriter.getStringAttr(foldedName));
+                        if (funcOp->hasAttr("qnode") and !foldedOp) {
+                            // Create the folded circuit function
+                            auto foldedCircuitAttr =
+                                getOrInsertFoldedCircuit(loc, rewriter, funcOp, foldingAlgorithm);
+                            func::FuncOp foldedOp =
+                                SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(
+                                    moduleOp, foldedCircuitAttr);
+                            std::vector<Value> args = {callOp.getArgOperands().begin(),
+                                                       callOp.getArgOperands().end()};
+                            args.push_back(currentFnFoldedOp.getArguments().back());
+                            rewriter.setInsertionPoint(callOp);
+                            rewriter.replaceOpWithNewOp<func::CallOp>(callOp, foldedOp, args);
+                        }
+                        else if (funcOp->hasAttr("qnode") and foldedOp) {
+                            rewriter.setInsertionPoint(callOp);
+                            std::vector<Value> args = {callOp.getArgOperands().begin(),
+                                                       callOp.getArgOperands().end()};
+                            args.push_back(currentFnFoldedOp.getArguments().back());
+                            rewriter.replaceOpWithNewOp<func::CallOp>(callOp, foldedOp, args);
+                        }
+                    });
+                    rewriter.restoreInsertionPoint(point);
```

Comment:
I find it difficult to follow the reasoning for restoring the insertion point here, inside of a conditional after we've modified it unconditionally.

I would suggest sticking to scope-based rewrite guards whenever possible.

Reply:
Removed by using PatternRewriter::InsertionGuard insertGuard(rewriter);